File: problem2.py
from ListNode import ListNode


class Solution(object):
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        add_one = False
        sum_node = ListNode(0)
        savehead = sum_node
        while l1 or l2 or add_one:
            sum_node.next = ListNode(0)
            sum_node = sum_node.next
            if l1:
                sum_node.val += l1.val
                l1 = l1.next
            if l2:
                sum_node.val += l2.val
                l2 = l2.next
            if add_one:
                sum_node.val += 1
                add_one = False
            if sum_node.val >= 10:
                sum_node.val -= 10
                add_one = True
        return savehead.next


b = Solution().addTwoNumbers(ListNode(2, next=ListNode(4, next=ListNode(3, next=ListNode(5)))),
                             ListNode(5, next=ListNode(6, next=ListNode(4))))
x = 1

File: java/problem2.java
class Solution2 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        boolean addOne = false;
        ListNode sumNode = new ListNode(), saveHead = sumNode;
        while (l1 != null || l2 != null || addOne) {
            sumNode.next = new ListNode(0);
            sumNode = sumNode.next;
            if (l1 != null) {
                sumNode.val += l1.val;
                l1 = l1.next;
            }
            if (l2 != null) {
                sumNode.val += l2.val;
                l2 = l2.next;
            }
            if (addOne) {
                sumNode.val += 1;
                addOne = false;
            }
            if (sumNode.val >= 10) {
                sumNode.val -= 10;
                addOne = true;
            }
        }
        return saveHead.next;
    }
}

public class problem2 {
    public static void main(String[] args) {
        Solution2 x = new Solution2();
        int[] a = { 2, 4, 3, 5 }, b = { 5, 6, 4 };
        x.addTwoNumbers(ListNode.createList(a), ListNode.createList(b)).printAll();
        System.out.println();
    }
}
